docs/proposals/SIL-Utilities.md

```diff
@@ -0,0 +1,106 @@
+# Utilities for SIL optimizer passes
+This document lists a set of SIL utilities to be used by the Swift implementation of the SIL optimizer.
+
+The purpose of this proposal is to do a better planning of what utilities we need in SIL optimizer passes.
+
+We want to avoid a situation like we have in the C++ SIL optimizer, where a huge amount of (partly) redundant utilities exist. Many of those utilities have overlapping functionality,  are difficult to discover and are poorly designed. We want to do better in the Swift SIL optimizer.
+
+## Basic Data-Structures
+#### `Stack`
+
+An allocation-free, array like data structure. To be used instead of `Swift.Array` wherever random-access is not needed.
+
+**Related C++ utilities:** `llvm::SmallVector`, `Stack`
+**Status:** done  
+
+#### `BasicBlockSet`
+An extremely efficient implementation of a set of basic blocks.
+
+**Related C++ utilities:** `BasicBlockSet`
+**Status:** done  
+
+#### `BasicBlockWorklist`
+To be used for all kind of basic-block work-list algorithms.
+
+**Uses:** `Stack`, `BasicBlockSet`  
+**Related C++ utilities:** `BasicBlockWorklist`
+**Status:** done  
+
+## Building SIL
+
+#### `Builder.insert(after:)`
```

Comment:
I think this should just be `builder = Builder(insert: .before(instruction)`

Reply:
There is of course a `Builder(at:)` (which inserts before). I didn't include that in the document because it's not really a utility, but a basic API.

What I meant here is the static function which inserts after the evaluation of an existing instruction. It takes a closure, which inserts in (multiple) successor blocks in case the insertion point is a terminator.

Reply:
Yes, that makes sense. It wasn't clear from the doc.